internals/secrethub/audit.go

```diff
@@ -34,27 +43,182 @@ func (cmd *AuditCommand) Register(r command.Registerer) {
 
 // Run prints all audit events for the given repository or secret.
 func (cmd *AuditCommand) Run() error {
+	cmd.beforeRun()
+	return cmd.run()
+}
+
+// beforeRun configures the command using the flag values.
+func (cmd *AuditCommand) beforeRun() {
+	cmd.timeFormatter = NewTimeFormatter(cmd.useTimestamps)
+}
+
+// Run prints all audit events for the given repository or secret.
+func (cmd *AuditCommand) run() error {
+	var auditTable auditTable
+	var iter secrethub.AuditEventIterator
+
 	repoPath, err := cmd.path.ToRepoPath()
 	if err == nil {
-		auditRepoCommand := AuditRepoCommand{
-			io:            cmd.io,
-			path:          repoPath,
-			useTimestamps: cmd.useTimestamps,
-			newClient:     cmd.newClient,
+		client, err := cmd.newClient()
+		if err != nil {
+			return err
+		}
+		tree, err := client.Dirs().GetTree(repoPath.GetDirPath().Value(), -1, false)
+		if err != nil {
+			return err
+		}
+
+		iter = client.Repos().EventIterator(repoPath.Value(), nil)
+		auditTable = newRepoAuditTable(tree, cmd.timeFormatter)
+	} else {
+		secretPath, err := cmd.path.ToSecretPath()
+		if err == nil {
+			if cmd.path.HasVersion() {
+				return ErrCannotAuditSecretVersion
+			}
+
+			client, err := cmd.newClient()
+			if err != nil {
+				return err
+			}
+
+			iter = client.Secrets().EventIterator(secretPath.Value(), nil)
+			auditTable = newSecretAuditTable(cmd.timeFormatter)
+		} else {
+			return ErrNoValidRepoOrSecretPath
 		}
-		return auditRepoCommand.Run()
 	}
 
-	secretPath, err := cmd.path.ToSecretPath()
-	if err == nil {
-		auditSecretCommand := AuditSecretCommand{
-			io:            cmd.io,
-			path:          secretPath,
-			useTimestamps: cmd.useTimestamps,
-			newClient:     cmd.newClient,
+	tabWriter := tabwriter.NewWriter(cmd.io.Stdout(), 0, 4, 4, ' ', 0)
+	header := strings.Join(auditTable.header(), "\t") + "\n"
+	fmt.Fprint(tabWriter, header)
+
+	// interactive mode is assumed, except when output is piped.
+	interactive := !cmd.io.Stdout().IsPiped()
+
+	i := 0
+	for {
+		i++
+		event, err := iter.Next()
+		if err == iterator.Done {
+			break
+		} else if err == api.ErrSecretNotFound {
+			// Check if we're attempting to audit a dir.
+			client, err := cmd.newClient()
+			if err != nil {
+				return err
+			}
+
+			_, err = client.Dirs().GetTree(api.DirPath(cmd.path).Value(), 1, false)
+			if err == nil {
+				return ErrCannotAuditDir
+			}
+			return api.ErrSecretNotFound
+		} else if err != nil {
+			return err
+		}
+
+		row, err := auditTable.row(event)
+		if err != nil {
+			return err
+		}
+
+		fmt.Fprint(tabWriter, strings.Join(row, "\t")+"\n")
+
+		if interactive && i == cmd.perPage {
+			err = tabWriter.Flush()
+			if err != nil {
+				return err
+			}
+			i = 0
+			fmt.Fprintln(cmd.io.Stdout(), "Press <ENTER> to show more results. Press <CTRL+C> to exit.")
```

Comment:
I guess this is because it is technically the easiest to implement. Do we want that to be the deciding factor on this?

Reply:
> Can't we just stop on a q and continue with any other key? Is also a bit more in line with other tools that allow a <space>.

Yes, that'd be nice.

> I guess this is because it is technically the easiest to implement.

Yes, that's indeed exactly why it's like this atm. I've tried to make it read single key presses, but couldn't get that to work quickly. Thought we would release this asap and change that later if we'd like to.

>Do we want that to be the deciding factor on this?

We can at any point add more options for stopping the process. imo, this can be a deciding factor in the short-term, but shouldn't be a big deciding factor in the long-term.

Feel free to pick this up now if you'd like to.

Reply:
Display-wise it is an MVP. But it works. So let's go with it.